### DISCUSSION TITLE
how to use, say, "reference only" now, that "None" model is gone via api?

### DISCUSSION THREAD

**Discussion by dermesut** [Mikubill/sd-webui-controlnet#2791] [Q&A] [answered]
title says it basically.
we had {"model_list": ["None", ...]} in http://127.0.0.1:7861/controlnet/model_list', but now "None" is gone.
it was convenient to have "None" in the model_list, since i could offer that in a drop-down of my app, if the user decided to go with a preprocessor like "reference_only". now i am basically forced to have "reference_only" combined with any of the entries in the model_list.
a solution could of course be, to code s.th. that recognizes the preprocessor selection, thus ignores what is set in the drop-down for the models. or s.th. similar.
but i wonder, why "None" had to be taken away, but even more so, if we could please get it back.

**Comment by huchenlei** [comment, marked as answer]
I don't think the API ever returns "None" in model list unless you are using a very old version of ControlNet. See model names impl here:
https://github.com/Mikubill/sd-webui-controlnet/blob/471f564651641d5d364b861a7f5088ca3d4d2b5d/scripts/global_state.py#L65-L89

In line 86, the None entry is filtered out from the name list. The function has not been touched for over a year now.

Please reopen if you can find a PR/commit that changes this behaviour recently.

**Comment by dermesut** [comment]
oh ok, thx. i can see that.
weird though, since i am not aware that i could have been using a very old version of controlnet.
i keep updating the controlnet extension as soon as i see a new version being indicated in the extensions-check in a1111.
i do that in all versions of a1111 that i have installed: 1.7.0, 1.8.0, 1.9.0 (which became 1.9.3 in the last days to hours).
and the version-indications in the controlnet units showed the correct updated version number.
since i have updated controlnet in all of the installations, i can not tell which commit hash was the last one that had "None" in the model_list, but it still was in 1.1.445. only the last one or two iteration of 1.1.445 "lost" the "None".
also, i do not "add anything" in my code... i just fetch the list via the given adress and list it in my app.
sorry, i have no idea, how this could be happening atm.

**Reply by huchenlei** [reply]
Can you find a good version where "None" is included in the model_list?

**Reply by dermesut** [reply]
i am sorry, that i am not experienced with how github/versioning etc. really works.
i just keep updating the version that is being offered in the check in a1111.
i can just tell, that i was using 1.1.445 (at least the first "iterations") with "None" included. and - of course - every version before that.
now, which commit that was, how to get that commit hash or a slightly older one.... and then install that one to test it... i have no idea.
i am very sorry. unless you can give me a quick tip on how to do that, i certainly don't want to steal your time with such novice stuff.

**Reply by dermesut** [reply]
so, i have been familiarizing myself a bit now.
i have tried diverse commits going back to 1.1.433, weeks old. git-cloning them always led to the same conclusion: there is no "None" being returned.
i also tried the same thing via the backup restore feature of a1111, again going back weeks.
same result. it's as though it never happened.
yet, yesterday i played around with "reference only" for a bit - i remember that well, because i rarely use that preprocessor - comfortably having a "None-model"..
this is very confusing to me - since every extension-update has been showing the evolving version number, giving me no reason to believe that i might actually have (very?) old stuff running, or at least bits of it.

well, now that i am in the same situation as the rest of the world apparently has been in for the past decade or two - having "None" not being returned - i'll figure out a way to live with that.

i am very sorry for wasting your time with such nonsense. thank you for your patience.
please close this, i'll mark your reply above as the answer.
